Improve our 'Creating widgets' section (#407)
napari/docs: docs/howtos/extending/magicgui.md

in building widgets. It is a general abstraction layer on GUI toolkit backends (like
Qt), with an emphasis on mapping Python types to widgets. This enables you to easily
create widgets using annotations.
If you require more extensibility, you can create your own widget `class` that
subclasses [`QtWidgets.QWidget`](https://doc.qt.io/qt-5/qwidget.html) or
{class}`magicgui.widgets.bases.Widget`.

This document will describe each widget creation method, in increasing order of
extensibility;

1. [{func}`@magicgui <magicgui.magicgui>` decorator](magicgui_decorator) - create a
   widget from a function and [`magicgui`](https://pyapp-kit.github.io/magicgui/).
   This is the simplest but least extensible option. It would suit one
   wishing to build a widget to simply run a function, with input widgets to select
   function parameters.

> 2. [](#magicgui-class-widgets) - subclass a

Reviewer: This renders with the hyphens in the name, is that by design?

Author: I ended up deleting the header this linked to. Should be fixed now

nbscreenshot(viewer, alt_text="A magicgui threshold widget")
```

```{note}
For a more complex example of a {func}`magicgui.magicgui` widget, see the
[gaussian blur example](https://pyapp-kit.github.io/magicgui/generated_examples/napari/napari_parameter_sweep/#napari-parameter-sweeps)
in the `magicgui` documentation.
```

See [](magicgui-parameter-annotations) for details on how to use `napari` types
to get information from the napari viewer into your widget. See
[](magicgui-return-annotations) for information how to use `napari` types to add
output to the napari viewer.

> Note all type annotations *require* that the resulting widget be added to a

Reviewer: What does this mean? Why does the widget have to be added to a viewer?

Author: This was definitely from Talley, and I wonder if it is a `magicgui` thing? Where the type annotation magic doesn't work until its added to a window? :shrug:

Author: Amended, hopefully that is okay

napari viewer.

To use these `magicgui` function widgets as plugin widget contributions,
see below.

(magicgui-plugin-widgets)=

### `magicgui` function widgets as plugin contributions

Recall [above](creating-widgets) that plugin
[widget contributions](widgets-contribution-guide) expects a `callable` that returns
a widget instance, whereas {meth}`~napari.qt.Window.add_dock_widget` expects an
*instance* of a widget. The {meth}`~napari.qt.Window.add_dock_widget` examples
above can be easily adapted to be plugin widgets by using
the {func}`@magic_factory <magicgui.magic_factory>` decorator instead of the
{func}`@magicgui <magicgui.magicgui>` decorator.

For example, the threshold widget [shown above](returning-napari-types-data)
could be provided as a napari plugin as follows:

```python
from magicgui import magic_factory

> from napari_plugin_engine import napari_hook_implementation

Reviewer: ```suggestion
```

@magic_factory(auto_call=True, threshold={'max': 2 ** 16})
def threshold(
    data: 'napari.types.ImageData', threshold: int
) -> 'napari.types.LabelsData':
    return (data > threshold).astype(int)
```

This function can now be added to the plugin manifest as a widget contribution.
See the [widget contribution guide](widgets-contribution-guide) for details.

:::{note}
{func}`@magic_factory <magicgui.magic_factory>` behaves very much like
{func}`functools.partial`: it returns a callable that "remembers" some or
all of the parameters required for a "future" call to {func}`magicgui.magicgui`.
The parameters provided to {func}`@magic_factory <magicgui.magic_factory>` can
also be overridden when creating a widget from a factory:

```python
@magic_factory(call_button=True)
def my_factory(x: int):
    ...

widget1 = my_factory()
widget2 = my_factory(call_button=False, x={'widget_type': 'Slider'})
```

:::

(magicgui-parameter-annotations)=

### Parameter annotations

The following `napari` types may be used as *parameter* type annotations in
`magicgui` functions or in the `annotation` argument of
{func}`magicgui.widgets.create_widget`. {func}`~magicgui.widgets.create_widget`

> can be used when adding an input widget to your widget class(TODO).

Reviewer: ```suggestion
can be used when adding an input widget to your widget class.
```

Accidental commit?

Author: Reminder to add a link

- {class}`napari.Viewer`

```{note}
When using {func}`~magicgui.widgets.create_widget` for a 'layer' type in a
[`QtWidgets.QWidget`](https://doc.qt.io/qt-5/qwidget.html) subclass,
you will need to manually connect the `reset_choices` of the resulting
{class}`~magicgui.widgets.ComboBox` (i.e., "dropdown menu") to layer events. This
is so it will synchronize with layer changes.
See the [`QWidget` example](#qwidget-class-widgets) for details.
```

The consequence of each type annotation is described below:

#### Annotating as a `Layer` subclass

Annotating a function parameter or setting `annotation` in `create_widget` to be a
{class}`~napari.layers.Layer` subclass (such as {class}`~napari.layers.Image` or

> {class}`~napari.layers.Points`), it will be rendered as a

Reviewer: ```suggestion
{class}`~napari.layers.Points`), will result in a 
```

viewer.window.add_dock_widget(my_widg)
```

As above, to turn this into a [plugin widget contribution](widgets-contribution-guide),
simply provide the class definition and add to the plugin manifest.

To build your widget from a 'blank slate', you can subclass
[`QtWidgets.QWidget`](https://doc.qt.io/qt-5/qwidget.html).
See [](#qwidget-class-widgets) for details.

### `QWidget` class widgets

To build your widget from a 'blank slate', subclass
[`QtWidgets.QWidget`](https://doc.qt.io/qt-5/qwidget.html).

In the following example, we create a button and a dropdown list.
The available choices for the dropdown list are the
current layers in the viewer. For this, we use
{func}`create_widget <magicgui.widgets.create_widget>` with the annotation
{attr}`napari.types.ImageData`.

Because the layer selection widget will be housed by a native `QWidget`
and not by a `magicgui` subclass (as with {func}`@magicgui <magicgui.magicgui>`

> decoratored functions and `magicgui` subclassees), we now need to

Reviewer: ```suggestion
decoratored functions and `magicgui` subclasses), we now need to
```
